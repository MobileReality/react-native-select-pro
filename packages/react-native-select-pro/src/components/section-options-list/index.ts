export { SectionOptionsList } from './section-options-list';
