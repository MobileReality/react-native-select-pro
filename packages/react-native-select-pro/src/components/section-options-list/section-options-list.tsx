import React, { memo, useRef } from 'react';
import isEqual from 'react-fast-compare';
import type { SectionListData } from 'react-native';
import { SectionList } from 'react-native';

import { ERRORS, isSectionSelected, logError } from '../../helpers';
import { getSectionLocation } from '../../helpers/get-section-location';
import type { OptionType, SectionOptionType } from '../../types';
import { NoOptions } from '../no-options';
import { SectionHeader } from '../section-header';

import type { SectionOptionsListProps } from './section-options-list.types';

export const SectionOptionsList = memo(
    <T,>({
        resolvedData,
        getItemLayout,
        renderItem,
        accessibilityState,
        selectedOption,
        scrollToSelectedOption,
        sectionListProps,
        disabled,
    }: SectionOptionsListProps<T>) => {
        const sectionOptionsListRef = useRef<SectionList>(null);

        const renderSectionHeader = <T,>(info: { section: SectionListData<T> }) => {
            const isSelected = isSectionSelected({
                title: info.section.title,
                selectedOptions: selectedOption as OptionType<T>[],
                sectionData: resolvedData as unknown as SectionOptionType<T>[],
            });

            return <SectionHeader title={info.section.title} isSelected={isSelected} />;
        };

        const scrollToIndex = () => {
            if (sectionOptionsListRef.current) {
                try {
                    sectionOptionsListRef.current.scrollToLocation({
                        ...getSectionLocation({
                            data: resolvedData,
                            selectedOption,
                            scrollToSelectedOption,
                        }),
                        animated: false,
                    });
                } catch {
                    logError(ERRORS.SCROLL_TO_LOCATION);
                }
            }
        };

        return (
            <SectionList
                testID="Options list"
                accessibilityLabel="Options list"
                accessibilityState={accessibilityState}
                bounces={false}
                keyboardShouldPersistTaps="handled"
                persistentScrollbar={true}
                ListEmptyComponent={<NoOptions />}
                scrollEnabled={!disabled}
                {...sectionListProps}
                ref={sectionOptionsListRef}
                renderSectionHeader={renderSectionHeader}
                sections={resolvedData}
                getItemLayout={getItemLayout}
                renderItem={renderItem}
                keyExtractor={({ value }) => value}
                onLayout={scrollToIndex}
            />
        );
    },
    isEqual,
);

SectionOptionsList.displayName = 'SectionOptionsList';
