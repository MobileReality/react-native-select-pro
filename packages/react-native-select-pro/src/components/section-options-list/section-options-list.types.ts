import type { SectionListProps } from 'react-native';
import type { FlatListProps } from 'react-native';

import type { State } from '../../state';
import type { OptionalToRequired, OptionType, SectionOptionType, SelectProps } from '../../types';

export type SectionOptionsListProps<T> = {
    resolvedData: SectionOptionType<T>[];
    renderItem: SectionListProps<OptionType<T>>['renderItem'];
    getItemLayout: SectionListProps<OptionType<T>>['getItemLayout'];
    accessibilityState: FlatListProps<OptionType<T>>['accessibilityState'];
} & OptionalToRequired<
    Pick<State<T>, 'selectedOption'> &
        Pick<SelectProps<T>, 'scrollToSelectedOption' | 'sectionListProps' | 'disabled'>
>;
