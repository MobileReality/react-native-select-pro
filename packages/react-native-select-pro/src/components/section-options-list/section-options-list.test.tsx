import React from 'react';
import { Text } from 'react-native';

import { SectionOptionsList } from './section-options-list';

describe('SectionOptionsList', () => {
    it('should generate SectionOptionsList snapshot', () => {
        const sections = [{ title: 'test', data: [{ label: 'test', value: 'test' }] }];
        const itemLayout = {
            length: 0,
            offset: 0,
            index: 0,
        };
        const sectionOptionsList = (
            <SectionOptionsList
                isOpened={false}
                optionsData={sections}
                noOptionsText={undefined}
                NoOptionsComponent={undefined}
                sectionListProps={undefined}
                sectionHeaderTextStyle={undefined}
                sectionHeaderContainerStyle={undefined}
                getItemLayout={() => itemLayout}
                renderItem={() => <Text>Test</Text>}
            />
        );
        expect(sectionOptionsList).toMatchSnapshot();
    });
});
