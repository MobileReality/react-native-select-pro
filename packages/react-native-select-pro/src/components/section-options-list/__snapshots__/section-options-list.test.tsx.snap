// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionOptionsList should generate SectionOptionsList snapshot 1`] = `
<RCTScrollView
  ListEmptyComponent={<NoOptions />}
  accessibilityLabel="Options list"
  accessibilityState={
    {
      "expanded": false,
    }
  }
  bounces={false}
  data={
    [
      {
        "data": [
          {
            "label": "test",
            "value": "test",
          },
        ],
        "title": "test",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="handled"
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  persistentScrollbar={true}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    [
      0,
    ]
  }
  testID="Options list"
>
  <View>
    <View
      style={null}
    >
      <View
        style={
          [
            {
              "backgroundColor": "#ffffff",
              "padding": 12,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#000000",
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          test
        </Text>
      </View>
    </View>
    <View
      style={null}
    >
      <Text>
        Test
      </Text>
    </View>
    <View
      style={null}
    />
  </View>
</RCTScrollView>
`;
